### PR TITLE
Add the ability to auto restart the last run when started after a crash

### PR DESCRIPTION
**Pull request by Veldrovive**
We are finding on the HPC that it is common for runs to crash and be auto-started again. This pull requests allows that auto restart to resume the previous run and load a checkpoint only when the run was auto restarted.

It also fixes a variable naming issue and unnecessary accelerator prepare call that was causing issues.

### CODE REVIEW

File: train_decoder.py [hunk 1]
```diff
@@ -289,9 +289,9 @@ def train(
     sample = 0
     samples_seen = 0
     val_sample = 0
-    step = lambda: int(trainer.step.item())
+    step = lambda: int(trainer.steps[0].item())
```
Comment (lucidrains) [inline train_decoder.py:292]: offered a helper method https://github.com/lucidrains/DALLE2-pytorch/commit/d7bc5fbedd85564f7050534d06a3d94da50e6c91#diff-617450527162fa367141dbf45e8b201673573479820af0ffc56ba93b7f70947fR537 this would only retrieve the number of training steps for the first unet

File: dalle2_pytorch/trackers.py [hunk 2]
```diff
@@ -398,20 +419,65 @@ def __init__(self, data_path: Optional[str] = DEFAULT_DATA_PATH, overwrite_data_
...
+                # Check if the logger is of the same type as the autoresume save
+                if auto_resume_dict['logger_type'] != self.logger.__class__.__name__:
+                    raise Exception(f'The logger type in the auto_resume file is {auto_resume_dict["logger_type"]} but the current logger is {self.logger.__class__.__name__}. Either use the original logger type, set `auto_resume` to `False`, or delete your existing tracker-data folder.')
+                # Then we are ready to override the logger with the autoresume save
+                self.logger.__dict__["resume"] = True
+                print(f"Upading {self.logger.__dict__} with {auto_resume_dict}")
+                self.logger.__dict__.update(auto_resume_dict)
+
+                return True
+        else:
```
Comment (lucidrains) [inline dalle2_pytorch/trackers.py:443]: more a stylistic critique, but one way for a cleaner function when you have all these branching nested if / else, is to put the edge cases near the top and do an early return

so in your function here

```python
if not self.auto_resume_path.exists():
  # log
  return False

if not self.logger.auto_resume:
  # log
  return False

# rest of the main logic loading from json + logging
return True
```

that way it is more readable and the edge cases are organized near the top

File: dalle2_pytorch/trackers.py [hunk 3]
```diff
@@ -398,20 +419,65 @@ def __init__(self, data_path: Optional[str] = DEFAULT_DATA_PATH, overwrite_data_
...
+            else:
+                # Otherwise we read the json into a dictionary will will override parts of logger.__dict__
+                with open(self.auto_resume_path, 'r') as f:
+                    auto_resume_dict = json.load(f)
+                # Check if the logger is of the same type as the autoresume save
+                if auto_resume_dict['logger_type'] != self.logger.__class__.__name__:
+                    raise Exception(f'The logger type in the auto_resume file is {auto_resume_dict["logger_type"]} but the current logger is {self.logger.__class__.__name__}. Either use the original logger type, set `auto_resume` to `False`, or delete your existing tracker-data folder.')
+                # Then we are ready to override the logger with the autoresume save
+                self.logger.__dict__["resume"] = True
+                print(f"Upading {self.logger.__dict__} with {auto_resume_dict}")
```
Comment (lucidrains) [inline dalle2_pytorch/trackers.py:439]: small typo

**Comment by lucidrains** [comment]
lgtm save for the training step being always fetched from the first unet! :100:

File: train_decoder.py (hunk 1 above)
Comment (Veldrovive) [reply train_decoder.py:292]: Will change to using that. I will also have to change the upsampling trainer pr to take advantage of this.

unet 0 is selected arbitrarily for no other reason than it does not matter which unet is chosen and 0 is always present. I am actually just going to remove this variable entirely and replace calls that use it with the helper function since that will be more reliably correct. It will still choose unet 0 though. I could also allow this to be configured, but since it only effects logging that does not seem neccesary.

Comment (lucidrains) [reply train_decoder.py:292]: oh ok, as long as you're aware!

**Comment by Veldrovive** [comment]
Made the changes and retested. Ignore the spelling mistake in the commit message.

File: train_decoder.py (hunk 1 above)
Comment (Veldrovive) [reply train_decoder.py]: I do wish there was a more satisfying way to do this, but it really is just an arbitrary choice and as long as the index is consistent between runs it is fine. The way it is handled in the upsampler training is that it takes the first training unet for the log step.

I guess we could go back to the original system of tracking the step separately from the unet step, but unet step is more informative so that doesn't seem like a good idea.
